<a href="https://colab.research.google.com/github/roysuraj115/Letsgrowmore-VIP-/blob/main/day_69.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


In [ ]:
df=pd.read_csv("diet.csv")

In [ ]:
df.head()

Foods Serving Size  Calories  Cholesterol (mg)  \
0      Frozen Broccoli    10 Oz Pkg      73.8               0.0   
1          Frozen Corn      1/2 Cup      72.2               0.0   
2  Raw Lettuce Iceberg       1 Leaf       2.6               0.0   
3       Baked Potatoes      1/2 Cup     171.5               0.0   
4                 Tofu    1/4 block      88.2               0.0   

   Total_Fat (g)  Sodium (mg)  Carbohydrates (g)  Dietary_Fiber (g)  \
0            0.8         68.2               13.6                8.5   
1            0.6          2.5               17.1                2.0   
2            0.0          1.8                0.4                0.3   
3            0.2         15.2               39.9                3.2   
4            5.5          8.1                2.2                1.4   

   Protein (g)  Vit_A (IU)  Vit_C (IU)  Calcium (mg)  Iron (mg)  \
0          8.0      5867.4       160.2         159.0        2.3   
1          2.5       106.6         5.2           3.3        0.3   
2          0.2        66.0         0.8           3.8        0.1   
3          3.7         0.0        15.6          22.7        4.3   
4          9.4        98.6         0.1         121.8        6.2   

   Price/Serving ($)  
0               0.48  
1               0.54  
2               0.06  
3               0.18  
4               0.93

In [ ]:
df.isnull().sum()

Foods                0
Serving Size         0
Calories             0
Cholesterol (mg)     0
Total_Fat (g)        0
Sodium (mg)          0
Carbohydrates (g)    0
Dietary_Fiber (g)    0
Protein (g)          0
Vit_A (IU)           0
Vit_C (IU)           0
Calcium (mg)         0
Iron (mg)            0
Price/Serving ($)    0
dtype: int64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17 entries, 0 to 16
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Foods              17 non-null     object 
 1   Serving Size       17 non-null     object 
 2   Calories           17 non-null     float64
 3   Cholesterol (mg)   17 non-null     float64
 4   Total_Fat (g)      17 non-null     float64
 5   Sodium (mg)        17 non-null     float64
 6   Carbohydrates (g)  17 non-null     float64
 7   Dietary_Fiber (g)  17 non-null     float64
 8   Protein (g)        17 non-null     float64
 9   Vit_A (IU)         17 non-null     float64
 10  Vit_C (IU)         17 non-null     float64
 11  Calcium (mg)       17 non-null     float64
 12  Iron (mg)          17 non-null     float64
 13  Price/Serving ($)  17 non-null     float64
dtypes: float64(12), object(2)
memory usage: 2.0+ KB


In [ ]:
!pip install pulp

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
import pulp as pl

In [ ]:
#Define Linear programming status here
max_nutrition = pl.LpProblem("diet_Problem",pl.LpMinimize)

In [ ]:
x=list(df["Foods"])
count=pd.Series(range(1,len(x)+1))
food=pd.Series(x)
foods=pd.concat([count,food],axis=1,keys=["S.no","Food_items"])
foods

S.no              Food_items
0      1         Frozen Broccoli
1      2             Frozen Corn
2      3     Raw Lettuce Iceberg
3      4          Baked Potatoes
4      5                    Tofu
5      6         Roasted Chicken
6      7      Spaghetti W/ Sauce
7      8               Raw Apple
8      9                  Banana
9     10             Wheat Bread
10    11             White Bread
11    12         Oatmeal Cookies
12    13               Apple Pie
13    14          Scrambled Eggs
14    15          Turkey Bologna
15    16        Beef Frankfurter
16    17  Chocolate Chip Cookies

Create a separate dictionary for each of these columns -'Price/Serving', 'Calories', 'Cholesterol (mg)', 'Total_Fat (g)', 'Sodium (mg)', 'Carbohydrates (g)', 'Dietary_Fiber (g)', 'Protein (g)', 'Vit_A (IU)', 'Vit_C (IU)', 'Calcium (mg)', and 'Iron (mg)'using zip()function

In [ ]:
cost=dict(zip(x,df["Price/Serving ($)"]))
Calories=dict(zip(x,df["Calories"]))
Cholesterol=dict(zip(x,df["Cholesterol (mg)"]))
Total_Fat=dict(zip(x,df["Total_Fat (g)"]))
Sodium=dict(zip(x,df["Sodium (mg)"]))
Carbohydrates=dict(zip(x,df["Carbohydrates (g)"]))
Dietary_Fiber=dict(zip(x,df["Dietary_Fiber (g)"]))
Protein=dict(zip(x,df["Protein (g)"]))
Vit_A=dict(zip(x,df["Vit_A (IU)"]))
Vit_C=dict(zip(x,df["Vit_C (IU)"]))
Calcium=dict(zip(x,df["Calcium (mg)"]))
Iron=dict(zip(x,df["Iron (mg)"]))

In [ ]:
cost

{'Frozen Broccoli': 0.48,
 'Frozen Corn': 0.54,
 'Raw Lettuce Iceberg': 0.06,
 ' Baked Potatoes': 0.18,
 'Tofu': 0.93,
 'Roasted Chicken': 2.52,
 'Spaghetti W/ Sauce': 2.34,
 'Raw Apple': 0.72,
 'Banana': 0.45,
 'Wheat Bread': 0.15,
 'White Bread': 0.18,
 'Oatmeal Cookies': 0.27,
 'Apple Pie': 0.48,
 'Scrambled Eggs': 0.33,
 'Turkey Bologna': 0.45,
 'Beef Frankfurter': 0.81,
 'Chocolate Chip Cookies': 0.09}

In [ ]:
food_vars=pl.LpVariable.dicts("Foods",x,lowBound=0,cat="Continuous")
food_vars

{'Frozen Broccoli': Foods_Frozen_Broccoli,
 'Frozen Corn': Foods_Frozen_Corn,
 'Raw Lettuce Iceberg': Foods_Raw_Lettuce_Iceberg,
 ' Baked Potatoes': Foods__Baked_Potatoes,
 'Tofu': Foods_Tofu,
 'Roasted Chicken': Foods_Roasted_Chicken,
 'Spaghetti W/ Sauce': Foods_Spaghetti_W__Sauce,
 'Raw Apple': Foods_Raw_Apple,
 'Banana': Foods_Banana,
 'Wheat Bread': Foods_Wheat_Bread,
 'White Bread': Foods_White_Bread,
 'Oatmeal Cookies': Foods_Oatmeal_Cookies,
 'Apple Pie': Foods_Apple_Pie,
 'Scrambled Eggs': Foods_Scrambled_Eggs,
 'Turkey Bologna': Foods_Turkey_Bologna,
 'Beef Frankfurter': Foods_Beef_Frankfurter,
 'Chocolate Chip Cookies': Foods_Chocolate_Chip_Cookies}

In [ ]:
max_nutrition+=pl.lpSum([cost[i]*food_vars[i] for i in x])
max_nutrition

diet_Problem:
MINIMIZE
0.48*Foods_Apple_Pie + 0.45*Foods_Banana + 0.81*Foods_Beef_Frankfurter + 0.09*Foods_Chocolate_Chip_Cookies + 0.48*Foods_Frozen_Broccoli + 0.54*Foods_Frozen_Corn + 0.27*Foods_Oatmeal_Cookies + 0.72*Foods_Raw_Apple + 0.06*Foods_Raw_Lettuce_Iceberg + 2.52*Foods_Roasted_Chicken + 0.33*Foods_Scrambled_Eggs + 2.34*Foods_Spaghetti_W__Sauce + 0.93*Foods_Tofu + 0.45*Foods_Turkey_Bologna + 0.15*Foods_Wheat_Bread + 0.18*Foods_White_Bread + 0.18*Foods__Baked_Potatoes + 0.0
VARIABLES
Foods_Apple_Pie Continuous
Foods_Banana Continuous
Foods_Beef_Frankfurter Continuous
Foods_Chocolate_Chip_Cookies Continuous
Foods_Frozen_Broccoli Continuous
Foods_Frozen_Corn Continuous
Foods_Oatmeal_Cookies Continuous
Foods_Raw_Apple Continuous
Foods_Raw_Lettuce_Iceberg Continuous
Foods_Roasted_Chicken Continuous
Foods_Scrambled_Eggs Continuous
Foods_Spaghetti_W__Sauce Continuous
Foods_Tofu Continuous
Foods_Turkey_Bologna Continuous
Foods_Wheat_Bread Continuous
Foods_White_Bread Continuous
Food

In [ ]:
pl.lpSum([food_vars[i]*Calories[i] for i in x])

67.2*Foods_Apple_Pie + 104.9*Foods_Banana + 141.8*Foods_Beef_Frankfurter + 78.1*Foods_Chocolate_Chip_Cookies + 73.8*Foods_Frozen_Broccoli + 72.2*Foods_Frozen_Corn + 81.0*Foods_Oatmeal_Cookies + 81.4*Foods_Raw_Apple + 2.6*Foods_Raw_Lettuce_Iceberg + 277.4*Foods_Roasted_Chicken + 99.6*Foods_Scrambled_Eggs + 358.2*Foods_Spaghetti_W__Sauce + 88.2*Foods_Tofu + 56.4*Foods_Turkey_Bologna + 65.0*Foods_Wheat_Bread + 65.0*Foods_White_Bread + 171.5*Foods__Baked_Potatoes + 0.0

In [ ]:
df[['Foods','Calories']]

Foods  Calories
0          Frozen Broccoli      73.8
1              Frozen Corn      72.2
2      Raw Lettuce Iceberg       2.6
3           Baked Potatoes     171.5
4                     Tofu      88.2
5          Roasted Chicken     277.4
6       Spaghetti W/ Sauce     358.2
7                Raw Apple      81.4
8                   Banana     104.9
9              Wheat Bread      65.0
10             White Bread      65.0
11         Oatmeal Cookies      81.0
12               Apple Pie      67.2
13          Scrambled Eggs      99.6
14          Turkey Bologna      56.4
15        Beef Frankfurter     141.8
16  Chocolate Chip Cookies      78.1

In [ ]:
max_nutrition += pl.lpSum([food_vars[x]*Calories[x] for x in food]) >= 800, "CaloriesMinimum"
max_nutrition += pl.lpSum([food_vars[x]*Calories[x] for x in food]) <= 1300, "CaloriesMaximum"
max_nutrition

diet_Problem:
MINIMIZE
0.48*Foods_Apple_Pie + 0.45*Foods_Banana + 0.81*Foods_Beef_Frankfurter + 0.09*Foods_Chocolate_Chip_Cookies + 0.48*Foods_Frozen_Broccoli + 0.54*Foods_Frozen_Corn + 0.27*Foods_Oatmeal_Cookies + 0.72*Foods_Raw_Apple + 0.06*Foods_Raw_Lettuce_Iceberg + 2.52*Foods_Roasted_Chicken + 0.33*Foods_Scrambled_Eggs + 2.34*Foods_Spaghetti_W__Sauce + 0.93*Foods_Tofu + 0.45*Foods_Turkey_Bologna + 0.15*Foods_Wheat_Bread + 0.18*Foods_White_Bread + 0.18*Foods__Baked_Potatoes + 0.0
SUBJECT TO
CaloriesMinimum: 67.2 Foods_Apple_Pie + 104.9 Foods_Banana
 + 141.8 Foods_Beef_Frankfurter + 78.1 Foods_Chocolate_Chip_Cookies
 + 73.8 Foods_Frozen_Broccoli + 72.2 Foods_Frozen_Corn
 + 81 Foods_Oatmeal_Cookies + 81.4 Foods_Raw_Apple
 + 2.6 Foods_Raw_Lettuce_Iceberg + 277.4 Foods_Roasted_Chicken
 + 99.6 Foods_Scrambled_Eggs + 358.2 Foods_Spaghetti_W__Sauce
 + 88.2 Foods_Tofu + 56.4 Foods_Turkey_Bologna + 65 Foods_Wheat_Bread
 + 65 Foods_White_Bread + 171.5 Foods__Baked_Potatoes >= 800

CaloriesMa

In [ ]:
#Carbohydrates' constraint
max_nutrition += pl.lpSum([food_vars[i]*Carbohydrates[i] for i in x]) >= 130, "CarbsMinimum"
max_nutrition += pl.lpSum([food_vars[i]*Carbohydrates[i] for i in x]) <= 200, "CarbsMaximum"

#Fat's constraint
max_nutrition += pl.lpSum([food_vars[i]*Total_Fat[i] for i in x]) >= 20, "FatsMinimum"
max_nutrition += pl.lpSum([food_vars[i]*Total_Fat[i] for i in x]) <= 50, "FatsMaximum"

#Protein's constraint
max_nutrition += pl.lpSum([food_vars[i]*Protein[i] for i in x]) >= 100, "ProteinsMinimum"
max_nutrition += pl.lpSum([food_vars[i]*Protein[i] for i in x]) <= 150, "ProteinsMaximum"

#Vit_A constraint
max_nutrition += pl.lpSum([food_vars[i]*Vit_A[i] for i in x]) >= 1000, "Vit_A_Minimum"
max_nutrition += pl.lpSum([food_vars[i]*Vit_A[i] for i in x]) <= 10000, "Vit_A_Maximum"

In [ ]:
max_nutrition.solve()


1

In [ ]:
max_nutrition.solver

In [ ]:
pl.LpStatus[max_nutrition.status]

'Optimal'

In [ ]:
# prob.variables()[1].varValue()

for var in max_nutrition.variables():
    print(f'Variable name: {var.name} , Variable value : {var.value()}\n')

print('\n')
print('*'*100)
print('\n')

#We can also see the slack variables of the constraints
for name, con in max_nutrition.constraints.items():
    print(f'constraint name:{name}, constraint value:{con.value()}\n')

print('*'*100)
print('\n')

## OBJECTIVE VALUE
print(f'OBJECTIVE VALUE IS: {round(max_nutrition.objective.value(),2)}')

Variable name: Foods_Apple_Pie , Variable value : 0.0

Variable name: Foods_Banana , Variable value : 0.0

Variable name: Foods_Beef_Frankfurter , Variable value : 0.0

Variable name: Foods_Chocolate_Chip_Cookies , Variable value : 0.0

Variable name: Foods_Frozen_Broccoli , Variable value : 1.4079473

Variable name: Foods_Frozen_Corn , Variable value : 0.0

Variable name: Foods_Oatmeal_Cookies , Variable value : 0.0

Variable name: Foods_Raw_Apple , Variable value : 0.0

Variable name: Foods_Raw_Lettuce_Iceberg , Variable value : 0.0

Variable name: Foods_Roasted_Chicken , Variable value : 1.2329408

Variable name: Foods_Scrambled_Eggs , Variable value : 4.0165704

Variable name: Foods_Spaghetti_W__Sauce , Variable value : 0.0

Variable name: Foods_Tofu , Variable value : 0.0

Variable name: Foods_Turkey_Bologna , Variable value : 0.0

Variable name: Foods_Wheat_Bread , Variable value : 0.0

Variable name: Foods_White_Bread , Variable value : 0.0

Variable name: Foods__Baked_Potatoes 

esult:** The optimal value of the Diet is $5.58. Which includes the following food items: 

--> 2.64 servings of baked potatoes

--> 4.02 servings of scrambled eggs

--> 1.23 servings of Roasted chicken

--> 1.41 servings of Frozen broccoli
